lib/receiver-queue.js

```diff
@@ -0,0 +1,42 @@
+const FIFO = require('fast-fifo')
+
+module.exports = class ReceiverQueue {
+  constructor () {
+    this.queue = new FIFO()
+    this.requests = new Map()
+  }
+
+  get length () {
+    return this.queue.length
+  }
+
+  push (req) {
+    this.queue.push(req)
+    if (req.id !== 0) this.requests.set(req.id, req)
+  }
+
+  shift () {
+    while (this.queue.length > 0) {
+      const req = this.queue.shift()
+      if (req.id !== 0) this.requests.delete(req.id)
+
+      if (req.block || req.hash || req.seek || req.upgrade) return req
```

Comment:
Just in case, if id is zero do we want to always return here?

Or zero id requests would always have one of those? Then it's fine as it is

Reply:
ya this is ok for now i think